[FLINK-4152] Allow re-registration of TMs at resource manager

tillrohrmann opened the pull request:
The `YarnFlinkResourceManager` does not allow the `JobManager` to re-register task managers have had been registered at the resource manager before. The consequence of the refusal is that the job manager rejects the registrations of these task managers. 

Such a scenario can happen if a `JobManager` loses leadership in an HA setting after it registered 
some TMs. The old behaviour was that the resource manager clears the list of registered workers and only accepts new registrations of task manager which have been started in a fresh container. However, in case that the previously registered TMs didn't die, they will try to reconnect to the new leader. The new leader will then ask the resource manager whether the TMs represent valid resources. Since the resource manager forgot about the already started containers, it rejects the TMs.

This PR changes the behaviour of the resource manager such that it can no longer reject TMs. Instead of being asked it will simply be informed about the registered TMs by the JM. If the TM happens to be running in a container which was started by the RM, then it will monitor this container. In case that this container dies, the RM will notify the JM about the death of the TM.

In that sense, the RM has no longer the authority to interfere with the JM-TM interactions and instead it is simply used as an additional monitoring service to detect dead TMs as a result of a failed container.

Furthermore, the PR adds a de-duplication method to filter out concurrent registration runs on the task manager. Before, it happened that a `RefusedRegistration` triggers a new registration run without cancelling the old registration run. This could lead to a massive amount of registration messages if the TaskManager's registration was refused multiple times.

The mechanism to de-duplicate `TriggerTaskManagerRegistration` works by assigning a registration run id which is changed whenever a new registration run is started. `TriggerTaskManagerRegistration` messages which have an outdated registration run id are then filtered out.

mxm commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -405,36 +374,13 @@ class JobManager(
-          }(context.dispatcher)
-
+          log.info(s"Register task manager $resourceId at the resource manager.")
+          rm ! decorateMessage(new RegisterResource(msg))
```
What if this message gets lost? This message needs to be delivered, otherwise the RM won't function properly. The RM may not run on the same node as the JobManager, making message loss even more likely.

tillrohrmann replied:
How would a lost message interfere with the behaviour of the RM? I thought that in the worst case the JM won't be notified about the death of a TM (if the container dies). However, this should also be recognized by the JM via the Akka's death watch.

mxm replied:
If containers die, then the ResourceManager will always be notified by Yarn and is able to pass this information to the JobManager. The advantage of ensuring that this message gets delivered upon TaskManager registration is that the ResourceManager can actually guarantee resources. On the other hand, if messages can be lost, the ResourceManager is just a tool to say "give me more", "give me less" with no actual guarantees how much you will get.

tillrohrmann replied:
If I'm not mistaken then there is hardly any difference between a registered worker and a container in launch. So in the current implementation it shouldn't matter much whether a container is in state "being launched" or "launched". Thus, it does not make much of a difference whether this message arrives or not.

Given that the `JobManager` does not yet use the RM to allocate new resources, it might actually be a good idea to regard the RM as a tool to notify the JM about TM failures. Everything else can be added once we actually need it.

mxm replied:
I just don't understand why you remove this functionality. It was not broken in any way. Of course, we can always add features later (that is true for any component) but it changes the original RM design. If we want to add monitoring of the pool size later on, we will have to re-add the proper registration at the RM.

tillrohrmann replied:
But it might be necessary to add the number of launched containers to the number of containers in launch in the `YarnFlinkResourceManager#getNumWorkersPendingRegistration`. Then the `checkWorkersPool` should behave correctly.

tillrohrmann replied:
Because it added complexity which is no longer needed. The current design still allows to change it later on. Therefore I don't see a problem in removing it. Just because we might need it in the future, is imho not a good reason to keep unused code around.

mxm commented:
Thank you for the pull request! Looking at the changes, it looks like it could have been broken up into two pull requests and jira issues. 1) Avoiding duplicate RegisterTaskManager messages 2) Changing core behavior of the ResourceManager.

Concerning 2, I would like to understand why it was necessary to change so much code. It seems like it would have sufficed to change one line of code (not clearing the bookkeeping on leader ship change). I'm not saying your changes don't make sense but I don't think they are backed by the original JIRA issue.

I'm not sure about the role change of the RM in this PR. The RM should be the authority for allocating new resources. If those resources are not properly reported back to the RM (e.g. message loss), the resource allocation won't work properly.

tillrohrmann commented:
It is true that the duplicate RegisterTaskManager messages fix could have been a separate PR. Since it surfaces as a symptom of the registration problem, I thought about fixing it in the same PR. The separation should be clear (TaskManager vs. RM changes).

The reason why I had to change more code than this single line is that the RM shouldn't interfere with the TaskManager registration to be on the safe side. As far as I've understood, the RM should tell you about a dying TM but it should not forbid the registration of a TM at the JM. I think that this conservative behaviour is more robust wrt the old behaviour and helps us to not overlook other corner cases.

Furthermore, some of the changes were necessary to make the YARN RM component more easily testable.

mxm commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkResourceManager.java:
```diff
@@ -78,6 +79,9 @@
 	private final Map<ResourceID, YarnContainerInLaunch> containersInLaunch;
 
+	/** The container where a TaskManager has been started and is running in */
+	private final Map<ResourceID, Container> containersLaunched;
```
Why this map? There is already a map in the base class which holds the registered resources.

tillrohrmann replied:
It is true that it holds the registered resources but it does not hold the launched containers. When a `JobManager` loses its leadership the list of registered workers will be cleared. In order to reconstruct the mapping `ResourceID --> Container`, you need this new map.

mxm replied:
What would be the drawback of not cleaning this list?

tillrohrmann replied:
You tell me, since you've authored this component. I guess there was a reason why you clear the list of registered resources when a JM loses its leadership.

tillrohrmann replied:
I basically introduced this map to not break the semantic contract of the `FlinkResourceManager` component.

mxm replied:
>You tell me, since you've authored this component. 

I think there is a misunderstanding. I did author this component but it is not mine. I know you spend some time working on this and I just want to understand your motives. I thought actually you had tried this fix since you mentioned it in the JIRA.

I think clearing the list is simply a bug and artifact of an old code design where the ResourceManager would be the central instance for TaskManager registration.

tillrohrmann replied:
I never said that it's yours but you've been much more involved in the design of this component than I was. Thus, I assumed that you know more than I do what the semantics of the `registeredWorkers` fields is. For example, does this field represent currently registered TMs at a JM or does this field represents the TMs which were registered at the RM by the JM?

I didn't know about it and couldn't deduce it from the code. Consequently, I chose the safest approach to leave it as it is and introduce a new container state in the `YarnFlinkResourceManager` implementation.

But if you think that clearing is just a bug and that the corresponding JavaDoc comment is outdated, then it might be a good option to change it.

mxm commented:
It was just a surprise to me that we're breaking the original ResourceManager role. If the ResourceManager doesn't get eventually notified about new TaskManagers, then it is not guaranteed anymore that the ResourceManager can fulfill pool size requests. The JobManager will have to deal with the pool size and the logic to ensure enough resources are available. Thus, the JobManager will be more concerned with managing resources.

tillrohrmann commented:
I think by imposing this contract we actually introduced problems we didn't have before. Thus, in order to remedy these problems for the upcoming release and go back a bit in the direction of the pre-RM age, I loosened the contract of the RM so that it can no longer reject TM registrations. This makes sense in my opinion, since we don't have a mean to shut down orphaned TMs anyway.

So in this version, the RM's task is to ensure that at least a predefined set of resources is allocated and to notify the JM about a TM death (not strictly mandatory).

What we could actually do is to also register orphaned TMs (or ones registered by a different RM). Then we wouldn't have the problem that we allocate too many resources for a JM.

mxm commented:
I don't feel particularly great about breaking the contract between JobManager and ResourceManager but it is a path we can go until we expand the ResourceManager capabilities. The figures here would have to be updated as well: https://issues.apache.org/jira/browse/FLINK-3543

Overall, the changes look good to me. I don't think we can reach consensus on the registration matter. As of now, I don't want to block release fixes. So +1 to merge if you feel like it.

tillrohrmann commented:
My main concern is actually that the JM-RM interaction is not well tested. Thus, I fear the more complex the code is the more possibility for mistakes there are. For example, I couldn't find a test where we test the `ReconnectResourceManager` functionality. Given that we're about to release 1.1 shortly, I would go for the simplest approach.

mxm commented:
+1 Let's get it out of the way for the 1.1 release.

tillrohrmann commented:
Hi @mxm, I've changed the implementation such that we no longer need the `containersLaunched` map in the `YarnFlinkResourceManager`. Instead we're not clearing the `registeredWorkers` map in the `FlinkResourceManager` when the `JobManager` loses leadership. Thus, the `registeredWorkers` field denotes the successfully started task managers (and the containers they are running in).

Additionally I reintroduced the reconnect resource manager functionality in the job manager. This should make sure that the resource manager is eventually notified about newly registered resources. In the current implementation, however, the resource manager will always accept the register resource messages. So only if the message gets lost and thus triggers a timeout exception, the reconnect resource manager message is sent.

Would be great if you could take another look at the changes.

mxm commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -405,36 +405,23 @@ class JobManager(
-
-      // ResourceManager knows about the resource, now let's try to register TaskManager
+      // ResourceManager is told about the resource, now let's try to register TaskManager
       if (instanceManager.isRegistered(taskManager)) {
```
Would it make sense to attempt registration of the task manager at the JobManager first and then, if successful, notify the ResourceManager? That's what the comment for `registeredWorkers` states now and I think it would make sense.

tillrohrmann replied:
Yes that's a good idea.

The `registeredWorkers` field can also contain TMs which were registered at a JM but now are not due to leadership loss or death watch. That's because the JM does not unregister resources at the moment. Maybe I should add this to the comment for `registeredWorkers`.

mxm replied:
We could also add an additional message to remove registered resources (upon Disconnect or Terminate). For most situations, the ResourceManager will probably detect a failed container anyways.

tillrohrmann replied:
Yes we could do this, but then we would have to introduce a kind of waiting state for the containers because the de-registration does not necessarily imply that the TM died. I think this would exceed the scope of this PR a little bit.

mxm replied:
I agree, it's out of the scope of this PR.

mxm commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/FlinkResourceManager.java:
```diff
@@ -113,7 +111,8 @@
 
-	/** The currently registered resources */
+	/** The resources which have been registered at the ResourceManager. This indicates the set
+	 * of containers where a TaskManager has been started and registered at a JobManager.  */
```
We should think about whether it denotes the startup and registration or simply the startup of a TaskManager process. This depends on whether we send the RegisterResource message before or after TaskManager registration at the JobManager.

tillrohrmann replied:
Yes as you've said it in another comment, it does not represent TMs which are currently registered but TMs which are started and have at least once been registered. For the sake of simplicity, we could rename the field into `startedWorkers`.

mxm replied:
Yes and we should change the comment to not state that they have been registered at the JobManager because currently they are registered at the ResourceManager before JobManager registration.

tillrohrmann replied:
Yes that is true. Will change it accordingly.

mxm commented:
Thank you for taking the time to revise some parts of the PR! I like the changes. The eventual resource registration has been re-introduced but simplified to only acknowledge the registration of TaskManagers at the JobManager. We could think about moving the acknowledgment of resources after the TaskManager registration at the JobManager. That would guarantee the full startup of the "worker". Actually, thinking about it again, it would break the semantics of the `registeredWorkers` map when the leader changes (because TaskManagers disconnect then). We may just leave it as it is.

tillrohrmann commented:
Hi @mxm, I've added the changes to rename `registeredWorkers` into `startedWorkers` and in the same wake I renamed the `RegisterResource` message into a `NotifyResourceStarted` message which reflects better its current usage. If you want, then you can take another look.

There is still some travis issue with the `YarnFlinkResourceManagerTest` case. Somehow the `YarnFlinkResourceManager` is not being started properly. I'm investigating.

mxm commented:
Hi @tillrohrmann! Your changes look good. I would say this is good to merge once we have fixed the issue with the `YarnFlinkResourceManagerTest` on Travis.

tillrohrmann commented:
At the moment I'm a little bit clueless why the test case errors occur. One test case error is probably related to concurrency issues in the `TestingLeaderRetrievalService`. I've fixed them but the error is still occurring in my local builds on Travis. Here the problem is that maven does not use the installed flink-runtim_2.10-tests.jar which contains the updated `TestingLeaderRetrievalService` but instead it downloads the the flink-runtime test jar from the snapshot repository which contains the code with the concurrency issue.

The other test issue I've seen is that the `flink-yarn-tests` sometimes fail because they contain no a
```
2016-07-22 12:01:31,308 ERROR akka.remote.EndpointWriter                                    - AssociationError [akka.tcp://flink@172.17.2.124:57922] <- [akka.tcp://flink@172.17.2.124:36043]: Error [Shut down address: akka.tcp://flink@172.17.2.124:36043] [
akka.remote.ShutDownAssociation: Shut down address: akka.tcp://flink@172.17.2.124:36043
Caused by: akka.remote.transport.Transport$InvalidAssociationException: The remote system terminated the association because it is shutting down.
]
```

in the logs. Since the test checks for exception as a word, it fails. I cannot yet explain why this association error now occurs but apparently didn't occur before.

tillrohrmann commented:
The latter problem apparently only occurred in my local branch because I set `akka.remote.log-remote-lifecycle-events=on` for debugging purposes. The former problem with the downloading of the flink-runtime test jar did not seem to occur here.

Let's see what the next Travis run says.

tillrohrmann commented:
The problem that maven downloads the flink-runtime_2.10-tests.jar from the snapshot repository instead of using the one from the local repository still remains.

mxm commented:
Thanks for investigating! Let me see if I can further debug the problem and find out why Maven doesn't use the installed `flink-runtime` artifact instead of the snapshot artifact.

tillrohrmann commented:
Thanks @mxm. Maybe this http://stackoverflow.com/questions/38021248/multi-module-maven-dependency-using-repository-not-local-on-travis-ci can help you.

mxm commented:
https://s3.amazonaws.com/archive.travis-ci.org/jobs/146612227/log.txt

After checking the logs again, I know why Maven isn't using the installed `flink-runtime` artifact. It tries to download the `flink-runtime_2.11` artifact when, in fact, it should try to download `flink-runtime-2.10`. Note, that the build is a Scala 2.11 build but this should only set the Scala version of dependencies but not change any Flink artifact names (we have a another script for that). When flink-runtime is built, it is built with Scala 2.11 although named flink-runtime_2.10. Still, `flink-yarn` attempts to download `flink-runtime_2.11`.

Trying to dig deeper, why this could be the case.

mxm commented on flink-yarn/pom.xml:
```diff
@@ -85,8 +85,20 @@ under the License.
+
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-runtime_${scala.binary.version}</artifactId>
```
Here's the culprit. This has to be `flink-runtime_2.10` because it is a Flink artifact. The `scala-2.11` property only affects Scala dependencies of Flink artifacts. The Flink artifacts names remain the sames unless changed via the `change-scala-version.sh` script.
 
If I remember correctly we don't use properties in artifact names because they are not resolved by Maven.

mxm commented:
Found the culprit. We must not use the `scala.binary.version` property in flink artifact names. The property is only for switching Scala dependencies. It is not resolved by Maven in artifact names during deployment which would give us artifact names of the form `flink-runtime_${scala.binary.version}`. The problem is that we set the property in the root pom which can only be deployed one time. Otherwise all artifacts would depend on Scala...a bit nasty.

The build failures occurred because Maven tried to use the `flink-runtime_2.11` dependency instead of `flink-runtime_2.10` (which is in fact the Scala 2.11 version if you use `-Dscala-2.11`). `flink-runtime_2.11` hadn't been installed previously, causing Maven to download it from the snapshot repository...

I think we're done with this PR after this. I'm going to run the changes on Travis and will merge to the master then.

tillrohrmann commented:
Thanks @mxm for resolving the issue. That was definitely a tricky one.

mxm commented:
Glad we figured it out :) Thanks for your PR!

EronWright commented:
@tillrohrmann @mxm I'll try to incorporate these changes into the [MesosFlinkResourceManager](https://github.com/EronWright/flink/blob/feature-MESOS-1984-T1/flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosFlinkResourceManager.java).   I deliberately kept the code as similar as possible to simplify future refactoring.

I too noticed the TM re-registration problem - that by removing items from registeredWorkers, a subsequent re-accept would fail.   Thanks for addressing it!
